cps/src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayAction.java

```diff
@@ -0,0 +1,242 @@
+        this.run = run;
+    }
+
+    @Override public String getDisplayName() {
+        return "Replay";
+    }
+
+    @Override public String getIconFileName() {
+        return isEnabled() ? "clock.png" : null;
+    }
+
+    @Override public String getUrlName() {
+        return isEnabled() ? "replay" : null;
+    }
+
+    private @CheckForNull CpsFlowExecution getExecution() {
+        FlowExecutionOwner owner = ((FlowExecutionOwner.Executable) run).asFlowExecutionOwner();
+        if (owner == null) {
+            return null;
+        }
+        FlowExecution exec = owner.getOrNull();
+        return exec instanceof CpsFlowExecution ? (CpsFlowExecution) exec : null;
+    }
+
+    /* accessible to Jelly */ public boolean isEnabled() {
+        if (!run.hasPermission(REPLAY)) {
+            return false;
+        }
+        CpsFlowExecution exec = getExecution();
+        if (exec == null) {
+            return false;
+        }
+        if (exec.isSandbox()) {
+            return true;
+        } else {
+            // Whole-script approval mode. Can we submit an arbitrary script right here?
+            return Jenkins.getActiveInstance().hasPermission(Jenkins.RUN_SCRIPTS);
+        }
+    }
+
+    /* accessible to Jelly */ public String getOriginalScript() {
+        CpsFlowExecution execution = getExecution();
+        return execution != null ? execution.getScript() : "???";
+    }
+
+    /* accessible to Jelly */ public Map<String,String> getOriginalLoadedScripts() {
+        CpsFlowExecution execution = getExecution();
+        return execution != null ? execution.getLoadedScripts() : /* ? */Collections.<String,String>emptyMap();
+    }
+
+    /* accessible to Jelly */ public Run getOwner() {
+        return run;
+    }
+
+    @RequirePOST
+    public void doRun(StaplerRequest req, StaplerResponse rsp) throws ServletException, IOException {
+        if (!isEnabled()) {
+            throw new AccessDeniedException("not allowed to replay"); // AccessDeniedException2 requires us to look up the specific Permission
+        }
+        JSONObject form = req.getSubmittedForm();
+        Map<String,String> otherScripts = new HashMap<String,String>();
+        for (String clazz : getOriginalLoadedScripts().keySet()) {
+            otherScripts.put(clazz, form.getString(clazz));
+        }
+        run(form.getString("mainScript"), otherScripts);
+        rsp.sendRedirect("../.."); // back to WorkflowJob; new build might not start instantly so cannot redirect to it
+    }
+
+    private static final Iterable<Class<? extends Action>> COPIED_ACTIONS = ImmutableList.of(
+        ParametersAction.class,
+        SCMRevisionAction.class
+    );
+
+    /** For whitebox testing. */
+    public @CheckForNull QueueTaskFuture/*<Run>*/ run(String script) {
+        return run(script, Collections.<String,String>emptyMap());
+    }
+
+    public @CheckForNull QueueTaskFuture/*<Run>*/ run(String script, Map<String,String> otherScripts) {
```

Comment:
> And we're ok with them being exposed to external use?

Yes, to the extent you can access `ReplayAction` from an API, it makes sense to call `run` on it.